hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,324 @@
+    case_node(c,cond,i, case) <-- conditional_node(c,cond),
+      for (i, case) in c.children(*cond).enumerate(),
+      if c.get_optype(case).is_case();
+
+    // inputs of conditional propagate into case nodes
+    out_wire_value(c, i_node, i_p, v) <--
+      case_node(c, cond, case_index, case),
+      io_node(c, case, i_node, IO::Input),
+      node_in_value_row(c, cond, in_row),
+      //in_wire_value(c, cond, cond_in_p, cond_in_v),
+      if let Some(conditional) = c.get_optype(*cond).as_conditional(),
+      if let Some(fields) = in_row.unpack_first(*case_index, conditional.sum_rows[*case_index].len()),
+      for (i_p, v) in (0..).map(OutgoingPort::from).zip(fields);
+
+    // outputs of case nodes propagate to outputs of conditional
+    out_wire_value(c, cond, OutgoingPort::from(o_p.index()), v) <--
+      case_node(c, cond, _, case),
+      io_node(c, case, o, IO::Output),
+      in_wire_value(c, o, o_p, v);
+
+    lattice case_reachable(C, Node, Node, bool);
+    case_reachable(c, cond, case, reachable) <-- case_node(c,cond,i,case),
+        in_wire_value(c, cond, IncomingPort::from(0), v),
+        let reachable = v.supports_tag(*i);
+
+}
+
+fn propagate_leaf_op<PV: AbstractValue>(
+    c: &impl DFContext<PV>,
+    n: Node,
+    ins: &[PV],
+) -> Option<ValueRow<PV>> {
+    match c.get_optype(n) {
+        // Handle basics here. I guess (given the current interface) we could allow
+        // DFContext to handle these but at the least we'd want these impls to be
+        // easily available for reuse.
+        op if op.cast::<MakeTuple>().is_some() => Some(ValueRow::from_iter([PV::new_variant(
+            0,
+            ins.iter().cloned(),
+        )])),
+        op if op.cast::<UnpackTuple>().is_some() => {
+            let [tup] = ins.iter().collect::<Vec<_>>().try_into().unwrap();
+            tup.variant_values(0, value_outputs(c.as_ref(), n).count())
+                .map(ValueRow::from_iter)
+        }
+        OpType::Tag(t) => Some(ValueRow::from_iter([PV::new_variant(
+            t.tag,
+            ins.iter().cloned(),
+        )])),
+        OpType::Input(_) | OpType::Output(_) => None, // handled by parent
+        // It'd be nice to convert these to [(IncomingPort, Value)] to pass to the context,
+        // thus keeping PartialValue hidden, but AbstractValues
+        // are not necessarily convertible to Value!
+        _ => c.interpret_leaf_op(n, ins).map(ValueRow::from_iter),
+    }
+}
+
+fn input_count(h: &impl HugrView, n: Node) -> usize {
+    h.signature(n)
+        .as_ref()
+        .map(Signature::input_count)
+        .unwrap_or(0)
+}
+
+fn value_inputs(h: &impl HugrView, n: Node) -> impl Iterator<Item = IncomingPort> + '_ {
+    h.in_value_types(n).map(|x| x.0)
+}
+
+fn value_outputs(h: &impl HugrView, n: Node) -> impl Iterator<Item = OutgoingPort> + '_ {
+    h.out_value_types(n).map(|x| x.0)
+}
+
+// Wrap a (known-length) row of values into a lattice. Perhaps could be part of partial_value.rs?
```

Comment:
I think this is an internal class (wrapping Vec into a lattice) and so rightly hidden, however, so currently I'm minded to keep it here - it's fairly short

Reply:
Moved into `value_row.rs` to make more space in `datalog.rs` for Machine to come back....(which then didn't take as much space as I expected because I left AnalysisResults behind in `results.rs`. Never mind.)

So, done.